File: akincraft/java/src/main/java/akincraft/entity/EntityManager.java
package akincraft.entity;

import org.joml.Vector3f;
import java.util.*;

/**
 * Manages all entities in the world - mobs, items, projectiles
 */
public class EntityManager {
    private List<Entity> entities = new ArrayList<>();
    private List<Entity> entitiesToRemove = new ArrayList<>();
    private Random random = new Random();
    
    public void update(float deltaTime) {
        for (Entity entity : entities) {
            entity.update(deltaTime);
        }
        
        // Remove dead entities
        entitiesToRemove.clear();
        for (Entity entity : entities) {
            if (entity.isDead()) {
                entitiesToRemove.add(entity);
            }
        }
        entities.removeAll(entitiesToRemove);
    }
    
    public void render() {
        for (Entity entity : entities) {
            entity.render();
        }
    }
    
    public void addEntity(Entity entity) {
        entities.add(entity);
    }
    
    public void removeEntity(Entity entity) {
        entities.remove(entity);
    }
    
    public List<Entity> getEntities() {
        return new ArrayList<>(entities);
    }
    
    public List<Entity> getEntitiesNear(Vector3f position, float radius) {
        List<Entity> nearbyEntities = new ArrayList<>();
        for (Entity entity : entities) {
            if (entity.getPosition().distance(position) < radius) {
                nearbyEntities.add(entity);
            }
        }
        return nearbyEntities;
    }
    
    public int getEntityCount() {
        return entities.size();
    }
}

File: akincraft/cpp/src/EntityManager.cpp
#include "EntityManager.h"
#include "Entity.h"
#include <algorithm>
#include <glm/gtc/type_ptr.hpp>

void EntityManager::update(float deltaTime) {
    for (auto& entity : entities) {
        entity->update(deltaTime);
    }
    
    // Remove dead entities
    entities.erase(
        std::remove_if(entities.begin(), entities.end(),
            [](const std::shared_ptr<Entity>& e) { return e->isDead_(); }),
        entities.end()
    );
}

void EntityManager::render() {
    for (auto& entity : entities) {
        entity->render();
    }
}

void EntityManager::addEntity(std::shared_ptr<Entity> entity) {
    entities.push_back(entity);
}

void EntityManager::removeEntity(std::shared_ptr<Entity> entity) {
    entities.erase(
        std::remove(entities.begin(), entities.end(), entity),
        entities.end()
    );
}

std::vector<std::shared_ptr<Entity>> EntityManager::getEntities() const {
    return entities;
}

std::vector<std::shared_ptr<Entity>> EntityManager::getEntitiesNear(
    const glm::vec3& pos, float radius) {
    std::vector<std::shared_ptr<Entity>> nearby;
    for (const auto& entity : entities) {
        if (glm::distance(entity->getPosition(), pos) < radius) {
            nearby.push_back(entity);
        }
    }
    return nearby;
}
